internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,162 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/keys"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+var _ storage.OpenFGADatastore = (*cachedDatastore)(nil)
+
+type cachedDatastore struct {
+	storage.OpenFGADatastore
+
+	cache        storage.InMemoryCache[any]
+	maxCacheSize int64
+	ttl          time.Duration
+}
+
+// NewCachedDatastore returns a wrapper over a datastore...
+func NewCachedDatastore(inner storage.OpenFGADatastore, cache storage.InMemoryCache[any], maxSize int64, ttl time.Duration) *cachedDatastore {
+	return &cachedDatastore{
+		OpenFGADatastore: inner,
+		cache:            cache,
+		maxCacheSize:     maxSize,
+		ttl:              ttl,
+	}
+}
+
+func (c *cachedDatastore) ReadUsersetTuples(ctx context.Context, store string, filter storage.ReadUsersetTuplesFilter, options storage.ReadUsersetTuplesOptions) (storage.TupleIterator, error) {
+	iter, err := c.OpenFGADatastore.ReadUsersetTuples(ctx, store, filter, options)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheKey := fmt.Sprintf("%s/%s#%s", store, filter.Object, filter.Relation)
```

Comment:
FYI, no action item: we have to be _super_ careful on what the cache keys are. If by mistake say we forgot to include `store` as cache key, we would giving the wrong tuples to some requests if there is a match on the other parts of the key.